src/server/pkg/storage/fileset/storage.go

```diff
@@ -287,6 +318,92 @@ func (s *Storage) WalkFileSet(ctx context.Context, prefix string, f func(string)
 	})
 }
 
+// SetTTL sets the time-to-live for the prefix. If the prefix contains multiple filesets
+// the ttl will be set on all of them and the soonest expiration time will be returned.
+// if no fileset is found SetTTL returns ErrNoFileSetFound
+func (s *Storage) SetTTL(ctx context.Context, prefix string, ttl time.Duration) (time.Time, error) {
+	// TODO: after we move the semantic paths to postrges the whole prefix can be set atomically.
+	var expiresAt *time.Time
+	if err := s.objC.Walk(ctx, applyPrefix(prefix), func(p string) error {
+		t, err := s.chunks.RenewReference(ctx, p, ttl)
+		if err != nil {
+			return err
+		}
+		if expiresAt == nil || t.Before(*expiresAt) {
+			expiresAt = &t
+		}
+		return nil
+	}); err != nil {
+		return time.Time{}, err
+	}
+	if expiresAt == nil {
+		return time.Time{}, ErrNoFileSetFound
+	}
+	return *expiresAt, nil
+}
+
+// RenewDuring ensures that none of the filesets under prefix expire by repeatedly calling SetTTL in a separate goroutine
+// until cb returns.
+// RenewDuring attempts to delete the contents of prefix after cb has returned
+func (s *Storage) RenewDuring(ctx context.Context, prefix string, cb func(context.Context) error) error {
+	const defaultTTL = 10 * time.Minute
+	eg, ctx2 := errgroup.WithContext(ctx)
+	ctx2, cancel := context.WithCancel(ctx2)
+	eg.Go(func() error {
+		var expiresAt time.Time
+		var err error
+		for {
+			expiresAt, err = s.SetTTL(ctx2, prefix, defaultTTL)
+			if err != nil && err != ErrNoFileSetFound {
+				return err
+			}
+			now := time.Now()
+			if err == nil && now.After(expiresAt) {
+				return errors.Errorf("fileset has already expired")
+			}
+			remaining := expiresAt.Sub(now)
+			if err := sleepCtx(ctx2, remaining/2); err != nil {
+				return nil // return nil on context cancelled
+			}
+		}
+	})
+	eg.Go(func() error {
+		defer cancel()
+		return cb(ctx2)
+	})
+	if err := eg.Wait(); err != nil {
+		return err
+	}
+	return s.Delete(ctx, prefix)
+}
+
+func (s *Storage) GetExpiresAt(ctx context.Context, prefix string) (time.Time, error) {
```

Comment:
Is this dead code? I cannot find a call site?

Reply:
Not used, but I think it's useful to expose this.  The API is pretty straightforward. I think it may be a good idea to have tests to ensure commits do not have a TTL associated with them.  Maybe even do that as part of a set of live integrity checks.

Reply:
I think we should remove it for the time being, because it would need to be changed anyway to accurately reflect the new model. I think most people would interpret this function as returning the expires at for a prefix, which is not how the garbage collection expires at works. Since it operates on independent paths, some paths may be removed much later than the returned time (and some paths might not even be removed at all if they don't have an expires at). I agree that integrity checks will be useful, but I don't think we should keep this around as is just because it could technically work to check if a commit has an expires at in any subfileset.